### PR TITLE
wr-153: make-email-case-insensitive

### PR DESCRIPTION
**Pull request by canterbery**

### CODE REVIEW

File: backend/src/db/migrations/20230906150722_make_users_email_index.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import { type Knex } from 'knex';
+
+async function up(knex: Knex): Promise<void> {
+  return await knex.raw(
+    'CREATE UNIQUE INDEX users_lowercase_email ON users (lower(email));',
+  );
+}
+
+async function down(knex: Knex): Promise<void> {
+  return await knex.raw('DROP index IF EXISTS users_lowercase_email');
```
Comment (andanf-e) [inline backend/src/db/migrations/20230906150722_make_users_email_index.ts]: Initially it has just unique constraint, so I believe it also should be moved back on rollback

File: backend/src/packages/users/user.repository.ts [hunk 2]
```diff
@@ -36,7 +36,7 @@ class UserRepository implements IRepository {
   public async findByEmail(email: string): Promise<UserEntity | null> {
     const user = await this.userModel
       .query()
-      .where({ email })
+      .whereRaw('LOWER(email)=?', [email.toLowerCase()])
```
Comment (andanf-e) [inline backend/src/packages/users/user.repository.ts]: As alternative
```suggestion
      .where('email', 'ilike', email)
```

File: backend/src/db/migrations/20230906150722_make_users_email_index.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
+import { type Knex } from 'knex';
+
+async function up(knex: Knex): Promise<void> {
+  return await knex.raw(
+    'CREATE UNIQUE INDEX users_lowercase_email ON users (lower(email));',
```
Comment (andanf-e) [inline backend/src/db/migrations/20230906150722_make_users_email_index.ts]: Let's also drop default constraint if it exists
`users_email_unique`

**Review by andanf-e — CHANGES_REQUESTED** [review]

File: backend/src/db/migrations/20230906150722_make_users_email_index.ts (hunk 1 above)
Comment (canterbery) [reply backend/src/db/migrations/20230906150722_make_users_email_index.ts]: I think DROP INDEX also removes UNIQUE constraint if I understood you correctly. 
image

Comment (andanf-e) [reply backend/src/db/migrations/20230906150722_make_users_email_index.ts]: I mean drop initial unique constraint on migration,, and turn it back on migration rollback

**Review by NazarKovtun1 — DISMISSED** [review]

File: backend/src/db/migrations/20230906150722_make_users_email_index.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import { type Knex } from 'knex';
+
+async function up(knex: Knex): Promise<void> {
+  return await knex.raw(
+    'CREATE UNIQUE INDEX IF NOT EXISTS users_lowercase_email ON users (lower(email)); alter table users drop constraint users_email_unique;',
+  );
```
Comment (andanf-e) [inline backend/src/db/migrations/20230906150722_make_users_email_index.ts]:
```suggestion
  await knex.raw(
    'CREATE UNIQUE INDEX IF NOT EXISTS users_lowercase_email ON users (LOWER(email));'
  );
  await knex.raw(
    'ALTER TABLE users DROP CONSTRAINT users_email_unique;'
  );
```

File: backend/src/db/migrations/20230906150722_make_users_email_index.ts [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+}
+
+async function down(knex: Knex): Promise<void> {
+  return await knex.raw(
+    'DROP index IF EXISTS users_lowercase_email; ALTER TABLE users ADD CONSTRAINT user_email_unique UNIQUE (email);',
+  );
```
Comment (andanf-e) [inline backend/src/db/migrations/20230906150722_make_users_email_index.ts]: Just to make queries more readable
```suggestion
  await knex.raw(
    'DROP INDEX IF EXISTS users_lowercase_email;'
  );
  await knex.raw(
    'ALTER TABLE users ADD CONSTRAINT user_email_unique UNIQUE (email);',
  );
```

**Review by mhrechkin — APPROVED** [review]

**Review by andanf-e — APPROVED** [review]

**Pull request merged**